atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -728,18 +735,32 @@ private Comparator<Cell> preserveInputRowOrder(List<Map.Entry<Cell, Value>> inpu
      * per non-empty row.
      */
     private Iterator<Map.Entry<byte[], RowColumnRangeIterator>> partitionByRow(
-            Iterator<Map.Entry<Cell, Value>> rawResults) {
+            Iterator<Map.Entry<Cell, Value>> rawResults, Iterator<byte[]> expectedRows) {
         PeekingIterator<Map.Entry<Cell, Value>> peekableRawResults = Iterators.peekingIterator(rawResults);
         return new AbstractIterator<Map.Entry<byte[], RowColumnRangeIterator>>() {
             byte[] prevRowName;
 
             @Override
             protected Map.Entry<byte[], RowColumnRangeIterator> computeNext() {
                 finishConsumingPreviousRow(peekableRawResults);
-                if (!peekableRawResults.hasNext()) {
+                if (!expectedRows.hasNext()) {
+                    Preconditions.checkState(
+                            !peekableRawResults.hasNext(),
+                            "Iterators are not consistent: there is some data returned by getRowsColumnRange()"
+                                    + " even when we expect no more rows. This is likely an AtlasDB product bug.",
+                            UnsafeArg.of("nextRawResult", peekableRawResults.peek()));
```

Comment:
I think we want the logic to be something like the following (also avoids allocating the arg in happy path):

```suggestion
                    if (peekableRawResults.hasNext()) {
                        throw new SafeIllegalStateException("Iterators are not consistent: there is some data returned by getRowsColumnRange()"
                                + " even when we expect no more rows. This is likely an AtlasDB product bug.",
                                UnsafeArg.of("nextRawResult", peekableRawResults.peek()));
                    }
```

Reply:
Yep, makes sense: I kind of added the unsafe-arg as an afterthought (and being a path that isn't supposed to happen didn't test it!)